cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "binary_ops.hpp"
+#include "operation.cuh"
+
+#include <cudf/binaryop.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/table/experimental/row_operators.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf::binops::compiled::detail {
+template <class T, class... Ts>
+inline constexpr bool is_any_v = std::disjunction<std::is_same<T, Ts>...>::value;
+
+template <typename DeviceComparator, class AdaptableUnaryFunction, class Iterator>
+struct device_comparison_functor {
+  using optional_iterator = thrust::transform_iterator<AdaptableUnaryFunction, Iterator>;
+  device_comparison_functor(optional_iterator const& optional_iter,
+                            bool const is_lhs_scalar,
+                            bool const is_rhs_scalar,
+                            DeviceComparator const& comparator)
+    : _optional_iter(optional_iter),
+      _is_lhs_scalar(is_lhs_scalar),
+      _is_rhs_scalar(is_rhs_scalar),
+      _comparator(comparator){};
+
+  bool __device__ operator()(size_type i)
+  {
+    return _optional_iter[i].has_value() &&
+           _comparator(cudf::experimental::row::lhs_index_type{_is_lhs_scalar ? 0 : i},
+                       cudf::experimental::row::rhs_index_type{_is_rhs_scalar ? 0 : i});
+  }
+
+  optional_iterator const _optional_iter;
+  bool const _is_lhs_scalar;
+  bool const _is_rhs_scalar;
+  DeviceComparator const _comparator;
+};
```

Comment:
Regarding the comment about this adding net lines of code: it's fine to skip the constructor and use [list initialization](https://en.cppreference.com/w/cpp/language/list_initialization) since the parameters are fairly self-contained.

Also this doesn't need so many template parameters, you can just write `template <typename OptionalIterator>` and not define `using optional_iterator = ...;`.

Finally, don't take iterators by reference. They're cheap to copy. See #11084. (But this is moot if we remove the constructor.)

I would try this:

```suggestion
template <typename OptionalIterator, typename DeviceComparator>
struct device_comparison_functor {
  bool __device__ operator()(size_type i)
  {
    return optional_iter[i].has_value() &&
           comparator(cudf::experimental::row::lhs_index_type{is_lhs_scalar ? 0 : i},
                       cudf::experimental::row::rhs_index_type{is_rhs_scalar ? 0 : i});
  }

  OptionalIterator const optional_iter;
  bool const is_lhs_scalar;
  bool const is_rhs_scalar;
  DeviceComparator const comparator;
};
```

No code changes should be needed at the calling sites.

Reply:
FYI: If you have public member variables then you don't need the constructor and can directly use initializer list for instantiating the struct. If the member variables are private then you must have a constructor.

Reply:
That's why we use `struct` instead of `class`. The members of a `struct` are public by default.

Reply:
I had tried this originally (implicit constructor from initializer list), but it has compilation failures.
```
/home/ryanlee/dev/rapids/cudf/cpp/src/binaryop/compiled/struct_binary_ops.cuh(88): error: no instance of constructor "device_comparison_functor" matches the argument list
            argument types are: (thrust::transform_iterator<cudf::detail::optional_accessor<__nv_bool, cudf::nullate::DYNAMIC>, cudf::column_device_view::count_it, thrust::use_default, thrust::use_default>, __nv_bool, __nv_bool, cudf::experimental::row::lexicographic::less_equivalent_comparator<cudf::experimental::row::lexicographic::strong_index_comparator_adapter<cudf::experimental::row::lexicographic::device_row_comparator<cudf::nullate::DYNAMIC, cudf::experimental::row::lexicographic::sorting_physical_element_comparator>>>)
```